flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchScheduler.java

```diff
@@ -232,8 +230,8 @@ protected void startSchedulingInternal() {
                     }
                 });
 
-        speculativeExecutionHandler.startSlowTaskDetector(
-                getExecutionGraph(), getMainThreadExecutor());
+        speculativeExecutionHandler.init(
```

Comment:
I prefer this to happen at the beginning of `startSchedulingInternal`, doing initialization before triggering scheduling.